### ISSUE TITLE
Feature request: A way to signal pods

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#24957]
This has come up a bunch of times in conversations.  The idea is that some external orchestration is being performed and the user needs to signal (SIGHUP usually, but arbitrary) pods.  Sometimes this is related to ConfigMap (#22368) and sometimes not.  This also can be used to "bounce" pods.

We can't currently signal across containers in a pod, so any sort of sidecar is out for now.

We can do it by `docker kill`, but something like `kubectl signal` is clumsy - it's not clearly an API operation (unless we revisit
'operation' constructs for managing imperative async actions).

Another issue is that a pod doesn't really exist - do we signal one container?  Which one?  Maybe all containers?  Or do we nominate a signalee in the pod spec?

**Comment by bprashanth** [comment]
We can just use config map as the kube-event-bus. It's what I was going to do for petset. 

I need to notify petset peers about changes in cluster membership. The idea is people will write "on-change" hook handlers, the petset will create a config map and write to it anytime it creates a new pet. The kubelet runs a handler exec/http-probe style, on every change to the config map.

I'd rather have a hook than an explicity signal because I don't need a proxy to listen for the signal and reload the pet. In fact to do this reliably I need a handler, because I need to poll till the pet shows up in DNS. 

I vastly prefer this to running a pid 1 that is cluster aware (autopilot strategy).

**Comment by thockin** [comment]
I opened this as a separate PR because I have been asked several times for
literal signals.  Asking people to do something instead of signals,
especially something kube-centric is a non-starter for people who have apps
that need signals.  We could ask them to bundle sidecars as bridges, but
because we can not signal across containers, they have to bundle into their
own containers.  That is gross to max.

**Comment by bprashanth** [comment]
You won't need a sidecar, you'll need a shell script that sends `pkill -HUP nginx`, right? the kubelet will exec-probe that script. This will require modifcation of pod spec but is more flexible imo.

**Comment by thockin** [comment]
You're missing the point.  There are other things people want to do that
involve sending a signal and "create a config map, modify your container to
include bash and pkill, modify your pod to mount the config map, write a
script that runs your process and then watches the configmap and signals
your process, and then run that instead of your real app" is a crappy
solution.  It's a kludgey workaround for lack of a real feature.

**Comment by bprashanth** [comment]
> write a script that runs your process and then watches the configmap and signals
your process, and then run that instead of your real app" 

That's not part of what i want. I explicitly *don't* want a long running process aware of the config map.

 >create a config map, modify your container to
include bash and pkill, modify your pod to mount the config map,

I think this is trivial. Not bash, sh or ash. Or http. The same things you need for a probe, or a post start hook, or pre stop hook. concepts we already have.

In short I want a way to tell the container about a reconfigure event and I don't want to teach all databases to handle a signal properly.

**Comment by bprashanth** [comment]
And I'm trying not to end up with 2 overlapping concepts when you can achieve one with the other. Somehow I dont think people will be against a probe like thing instead of a straight signal.

**Comment by thockin** [comment]
It feels like a hack to me.  I'd be fine with notifiers, and with notifiers
attached to configmap changes, and with notifiers that included "send a
signal".  None of that obviates the utility of "i want to signal my pods"

**Comment by bprashanth** [comment]
Hmm, my feature request also doesn't require a config map, it's the common case. Maybe a notifier of type=probe,exec,signal?

**Comment by smarterclayton** [comment]
Regarding signals and restarts, is kill the right signal, or do users
want to restart the pod itself?  I fee like the granularity on signals
is processes (of which container is a relatively good proxy), but the
granularity on restart / exec actions is containers or pods.

**Comment by thockin** [comment]
I could see an argument for "restart" as a verb on pods (albeit with he
same issues around imperatives), distinct from signal, though I assumed
signalling a pod meant to signal all containers or a self-nominated signal
receiver container.

**Comment by smarterclayton** [comment]
Is "restart" a synthetic signal then?  If we add exec it's a little
wierd because we would potentially have to secure exec differently
than signal (since it could be running with higher priviliges) - do
pods authors define signals the container supports (which is more like
invokable hooks), or are they arbitrary (can I find crazy posix signal
X that causes the machine to crash)?

**Comment by bprashanth** [comment]
I'm still kind of stuck at the point where we're special casing a specific form of linux IPC. I really want a cluster notification system, the last hop of which is some implementation of a node local ipc, i.e, say i have a pod with:

```
notification:
  exec: /on-change.sh
  stdin: ""
```

and I write "foo" to stdin, the kubelet will deliver that at least once, to the one container with the hook. I then observe generation number and know that my container has received `echo foo | on-change.sh`. Now say i want to send a signal, I'd just do:
```
notification:
  signal: posix number??
```

or
```
notification:
  exec: "sighup.sh"
  stdin: ""
```
And I still write "" to the stdin field, and poll on generation.

Is that a different feature request?

**Comment by smarterclayton** [comment]
I kind of agree signal is a subset of "pod notification".  The security
aspects of exec make it something I'd prefer to require the pod author to
define, in which case it's a named hook / notification.

**Comment by smarterclayton** [comment]
I would like to have a concrete way to force and guarantee a restart of a
pod from the outside without hacking.  One option that is not signals is a
monotonic value on the pod that a user can update (essentially,
generation).  That's more "whole pod" signaling.

**Comment by bprashanth** [comment]
So all containers BUT pause need to restart? (otherwise why not just delete the pod?)

**Comment by eghobo** [comment]
it should be optional and probably support rollout restart, it's too dangerous to restart everything right a way.

**Comment by bprashanth** [comment]
> it should be optional and probably support rollout restart,

If it's too dangerous to restart containers *in* a pod at the same time, put them in different pods/services? Everything else in the system will treat the pod as a unit, so there's probably no avoiding this. Or are you asking for a dependency chain of restarts on containers in a pod?

**Comment by smarterclayton** [comment]
Classic cases:

* service registers itself on start with a system, that systems failed
* server has caches that need to be flushed because of a failure elsewhere
* server is flaking in a way that liveness probe doesn't/can't catch
* debugging a pod on a particular node
* database that needs a restart after a schema change

In all of these cases the invariant enforced by pod deletion is excessive.

**Comment by eghobo** [comment]
@bprashanth: for example we have replication controller which has N of replicas, if we are going to restart all of them at the same time (e.g. we rollout new config change) we will affect customer traffic.

**Comment by therc** [comment]
More use cases: some software can be told to increase logging or debugging verbosity with SIGUSR1 and SIGUSR2, to rotate logs, open diagnostic ports, etc.

**Comment by pmorie** [comment]
I also view this as notification.  I would prefer to expose an intentional API instead of something in terms of POSIX signals.  I would take two actions to start with:

1.  `bump` - signal the container somehow that a reload should happen of any config
2.  `restart` - what it says on the tin

The problems with an intentional API  are:

1.  The ways of carrying out intent are not going to be universal across platforms; we need to define them for each platform we support; we can say that on Linux, `bump` means SIGHUP
2.  As others have commented, there will be some overlap with other areas like the use of `SIGUSR1` and `SIGUSR2` - where do we draw the line between rich APIs for specific platforms versus more abstract (but harder to implement universally) APIs?

**Comment by pmorie** [comment]
Also, for the record, seems plausible that we would constrain the list of signals you could send in an API using `PodSecurityPolicy`.

**Comment by nhlfr** [comment]
I have no strong preference about "intentional API vs POSIX" problem. The only proposition I'm opposed is using any king of exec or scripts as a workaround. Docker has `docker kill --signal`, rkt has the issue where the idea of supporting custom signals is considered coreos/rkt#1496. I'd prefer to go with these features.

@pmorie sorry for my ignorance, but what exact non-Linux platforms we have to care about in k8s? I had no luck with googling/greping that information.

**Comment by smarterclayton** [comment]
There is ongoing work to bring Kubernetes to Windows.  We need to be
aware of how we would address differences in those mappings.

**Comment by bprashanth** [comment]
For the record we should be able to exec scripts from sidecars for notification delivery with shared pidnamespaces, which is coming in docker 1.12 apparently (https://github.com/kubernetes/kubernetes/issues/1615)

**Comment by nhlfr** [comment]
@smarterclayton OK, so in that case intentional API sounds better.

So, for now we're going to implement the `bump` and `restart` actions as

1. API actions for pod
2. Actions which may be triggered by event bus

?

Also, as far as I understand, this event bus is something which we have to implement? Sorry if I'm missing some already existing feature.

If I understand correctly the API part, then I'm eager to implement this.

**Comment by pmorie** [comment]
@nhlfr For a change like this the best way to start is a proposal document that soecifies the yse-cases and proposes the API changes. Implementation concerns come later, but to answer your question, we already have a watch emchnaism in the API that kubelet uses to watch for changes in pods, and can be used with any API type.

**Comment by thockin** [comment]
Sorry, I walked away from this for a while.

I *dislike* an opinionated "bump" operation because that isn't what I asked for.  I have been asked repeatedly for "I want to send a signal".  Yes, I know that signals are a POSIXism and Windows support is approximately inevitable.  That doesn't mean that I am going to change my app from expecting SIGUSR1 to now expect SIGHUP, even if SIGHUP is more obviously correct.  We don't get to choose that.

I would be OK to cage this in terms of a notifier API, where the operation at the API level is "trigger this pod's XYZ notifier" and the pod defines XYZ as "signal: SIGUSR1" or something.  That's clean and extensible, and is actually pretty close to what Google has internally, too.  But at the end of the day a signal of the user's choosing is delivered.  Obvious extensions are an exec or an HTTP GET or POST.

If that's all you were saying, @bprashanth, then I am sorry for not getting it before :)

**Comment by smarterclayton** [comment]
I think both @bprashanth and I were suggesting pod or container defined named hooks at various points that were then treated as one or more "signals", and that's not terribly different from what's being proposed.  If a pod author wanted to create:

```
notification:
  - name: bump
    signal: SIGUSR1
```

that's both intentional and extensible, as you note.

I don't think there's an "event bus" per se, but we do need to talk about delivery guarantees.  As a client author I should not have to write a retry loop around notifications.

**Comment by thockin** [comment]
I agree with this last message in its entirety.  Sorry for being obtuse
before.  This would satisfy what I understand as requirements.

**Comment by smarterclayton** [comment]
Not obtuse at all - I think you helped crystalize it.  I had been
neglecting this as well.

**Comment by nhlfr** [comment]
@smarterclayton @thockin So, assuming that we'll define a `bump` notification in the definition of pod, then we'll have some API endpoint in which the name of notification will be provided? Something like _/v1/api/namespace/<ns>/pods/<pod>/notify_? Sorry if I misunderstood something.

**Comment by thockin-cc** [comment]
Something like that, yeah.  The persistence, acknowledgement, and
guarantees of the API are what become tricky

**Comment by smarterclayton** [comment]
There's a question about whether signal delivery must be guaranteed -
whether even in the presence of a Kubelet crash we will be guaranteed to
execute it.  That might require storage of an actual field / object in etcd
(either on pod or off) that the kubelet consumes.

**Comment by nhlfr** [comment]
Can we consider the signal delivery guarantee as one of the future improvements? Of course it would be great to store information about signals in etcd and ensure that kubelet crash will not prevent signal sending, but IMO we should focus on that after the initial implementation. What do you think?

Of course I'm going to implement that etcd part as well, but I just want to keep it as a separate step of development.

**Comment by kfox1111** [comment]
What about a more generic action kind of thing.... Then users could register other actions that they may want to preform on containers too.

something like:
kubecli action [podname] <-c optional_container_name> [actionname] <optional args>

and in a containerSpec:

actions:
  signal:
     exec:
         command: 
           - kill
           - Ref: args
           - 1
   debuglogging:
     exec:
        command:
          - /debug.sh
          - Ref: args

where debug.sh might be like:
   [ $1 == "on" ] && echo DEBUG=255 > /etc/myapp/debug || rm -f /etc/myapp/debug
   kill -SIGHUP 1

and can be used like:
kubectl action mypod -c mycont debuglogging on
kubectl action mypod -c mycont debuglogging off
kubectl action foo signal -SIGHUP

By default we could add the signal handler above so users don't have to register that one, but can easily override it.

**Comment by thockin-cc** [comment]
I don't think the extended actions are very attractive right now.  Given
how complex it is to do imperative well, and the pushback we're getting
(rightly), I think less is more.

Can we instead find ways to NOT do this (I know I opened it, so I am sort
of shooting myself down :)?  It's such a can of worms, and I am worried
that we keep adding floors to the building, when we really should be
reinforcing the foundation...

**Comment by bprashanth** [comment]
just watch your own pod annotations and hup the peer container once we upgrade docker to 1.12

**Comment by nhlfr** [comment]
@kfox1111 what is the point of your idea? You can already exec commands in the containers, so this functionality already exists.

The reason why we want to "hide" signals or HTTP requests under the concept of "notification" we're discussing currently, is that:

* we don't want to allow to send every signal possible everywhere, we rather prefer to define signals which can be used per container
* in case of HTTP reqs we have to provide defined list of possible HTTP notifications

And BTW, I think the level of abstraction we agreed on currently is enough and I'd be against any more "meta", "abstract" proposals here.

**Comment by kfox1111** [comment]
@nhlfr the idea being the op operating the containers can be different then the dev that wrote the deployments and is an abi in that relationship so the op doesnt need to know the details of performing some action. Yes, you could do it with an exec, and I think maybe that should be how its implemented. Exec is more of an unstructured back door though.

**Comment by ChipmunkV** [comment]
So in the end, how to make it send SIGHUP when the configuration changes?

**Comment by thockin** [comment]
As of now there is no supported API for this.  I still think it is worth
doing something here, but what exactly is complicated, and I have not had
time to think more deeply on it.

**Comment by nhlfr** [comment]
@ChipmunkV `kubectl exec -ti name-of-pod /bin/kill -SIGHUP 1`, assuming that your PID 1 in the container is the main process you want to kill.

**Comment by ChipmunkV** [comment]
> @ChipmunkV kubectl exec -ti name-of-pod /bin/kill -SIGHUP 1, assuming that your PID 1 in the container is the process you want to kill.

@nhlfr okay, but my hypothetical system has just two buttons "upgrade application to the new version" and "upgrade config file of the application to the new version". Is there some preferred place for a hook that can execute that command when the kubernetes trigger config file change?

**Comment by kargakis** [comment]
> @nhlfr okay, but my hypothetical system has just two buttons "upgrade application to the new version" and "upgrade config file of the application to the new version". Is there some preferred place for a hook that can execute that command when the kubernetes trigger config file change?

Related: https://github.com/kubernetes/kubernetes/issues/22368

**Comment by discordianfish** [comment]
I came across this issue long time ago and didn't bothered much because I thought it would be a no brainer and I'm a bit surprised that people question whether this is a necessary feature.

Signals are one of the most fundamental tools for me as an operator to talk to my processes. While I agree that there are nicer ways to do IPCs, it's not that everyone will consolidate around these anytime soon. In the meanwhile I need signal.

More specifically, what am I suppose to do if I need to HUP a process? Or send USR1 to increase log level temporary (which, again might not be the ideal IPC but commonly used). The only option I have right now is sshing into the hosts, figuring out the PID and send a signal from there. Or i need to include `kill` and possible dependencies into my containers, so I can use exec.. Unless there is a way to run a container in the same pid namespace as another container or something.

Maybe I miss something, but most of this issue here seems to discuss pros and cons of various notification systems which is orthogonal to supporting the most fundamental, most established notification process we have - signals.

**Comment by thockin** [comment]
It's not that we (I) don't want this.  I do.  But building it in a way that
is robust enough to rely on involves some real trail-blazing.  we need a
model for imperative actions, a spec around delivery guarantees, and some
serious tests.  As far as I know, nobody is stepping up to implement this,
and I'm not sure we have bandwidth to sponsor it right now, even.

I think we'd love to hear alternative ideas on how to build it in a way
that was less onerous, but still useful...

**Comment by discordianfish** [comment]
Ok cool! The history here sounded like there wasn't even an agreement that this is needed. Is that something that could piggyback on jobs? But to be fair, I'm not familiar enough with the k8s internals..

With #1615 I could at least have a 'job' sidecar for sending signals.

**Comment by thockin** [comment]
The problem is fundamentally that a signal probably means "do this now" and
"guarantee it go done".  For example, if the guarantee was that `kubectl
signal mypod hup` would deliver the signal within 1 minute, but only
delivered the signal 75% of the time, could you rely on it?  I couldn't.

To make it really reliable we need to design a new thing in the API that
has no precedent and is fairly different in meaning than anything else
(therefore will need some serious contemplation).  I like the idea of
designing it, and I think it does solve real problems, but I also think
there are existing workarounds and there are other things that are way more
important to spend both design time and complexity budget on.

That said, if someone wanted to prototype things, with no promise of
eventual merge, it might be a fun way to learn the guts of Kubernetes.

**Comment by scippio** [comment]
any news? :thinking:

**Comment by jstoja** [comment]
From my understanding, this feature is stalled until a proper design is layed out. I assume this is a pretty complex piece and as explained by @thockin it needs to be robust and reliable.
Is there any label to add for technical design needs?

**Comment by thockin** [comment]
There's no generic tag for "design needed".  The "feature request" in the title says that, to me.

**Comment by chrishiestand** [comment]
Does this PR solve the issue for some/most use-cases? e.g. afaik, a signal sidecar should be possible
 https://github.com/kubernetes/kubernetes/pull/45236

**Comment by nhlfr** [comment]
@chrishiestand Depends whether you ask about hacky solution you can use here and now for killing containers/processes, or whether you are concerned about designing a proper API for killing containers.

Well, if you have shared PID namepsace, you can add a container with `kill` binary to the pod (which is continuously tunning, i.e. `sleep` in the loop), execute it (`kubectl exec`) and kill any process from the pod you want. This may be a good solution for today.

But I don't think it really helps in the discussion we had about implementing an API.

**Comment by jeffrey4l** [comment]
any news for this?

use `kubectl exec` is not convenience when there  are lots of pod replicates.

**Comment by verb** [comment]
It's not an API for signaling, but [signaling across containers in a pod](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/) is available as an alpha feature in 1.10, which could enable sidecars.

You can follow along and give feedback in kubernetes/features#495

**Comment by subos2008** [comment]
This issue was linked from https://github.com/kubernetes/kubernetes/issues/29761 (and the source issue closed).

That original issue is saying we want to restart containers when secrets change. I see that as a simpler case than the more complex issue of signalling here. Secrets can be set in the environment, the viewable environment is in some ways set on process startup, so if a secret used to set an environment variable is changed then you could say that container is flagged for re-creation.

How is rolling updates on image handling done? Can we use similar logic to flag pods with out of date environments to need 'updating' and restart them in a safe rolling manner?

Would others agree splitting the issues of "I want to send custom signals to a pod" and "outdated pods getting replaced" makes ... well, means we might move some of the less complex use cases forward?

**Comment by alitoufighi** [comment]
Regarding comparisons of doing a rolling restart instead of just sending signals, in my case, I have a Prometheus deployment that uses a volume to persist the time series, so that its data is huge. When restarting, it takes about 5 minutes to read the tsdb to become available.

But if I send a `SIGHUP` signal after changing its config in ConfigMap, it only reloads the config and is available without any downtime.

Currently I am using `kubectl exec -it $PROMETHEUS_POD -- /bin/sh -c "/bin/pkill -HUP prometheus"` as a workaround, but I think there must be a native Kubernetes approach to do this.

**Comment by joemiller** [comment]
@alinbalutoiu You could use a sidecar container that watches the mounted configmap for changes and sends a signal to prometheus. That's the best/closest to native way I'm aware of.

in the pod spec: `  shareProcessNamespace: true`

A bash script with while loop around inotify
```bash
#!/bin/bash
# requires these rpm packages on fedora/centos flavors:
# - procps-ng
# - inotify-tools
# or, requires these deb packages on debian/ubuntu flavors:
# - procps
# - inotify-tools

set -eu -o pipefail

WATCH_FILE="/configmap-mount/prometheus.config"
INOTIFYWAIT_OPTS="${INOTIFYWAIT_OPTS:-"-e modify -e delete -e delete_self"}"

main() {
  while true; do
    while [[ ! -f "$WATCH_FILE" ]]; do
      echo "Waiting for $WATCH_FILE to appear ..."
      sleep 1
    done

    echo "Waiting for '$WATCH_FILE' to be deleted or modified..."
    # shellcheck disable=SC2086
    if ! inotifywait $INOTIFYWAIT_OPTS "$WATCH_FILE"; then
      echo "WARNING: inotifywait exited with code $?"
    fi
    # small grace period before sending SIGHUP:
    sleep 1

    echo "sending SIGHUP to prometheus"
    if ! pkill -HUP prometheus; then
      echo "WARNING: 'pkill' exited with code $?"
    fi
  done
}
main "$@"
```

inotifywait will block until any changes to the $WATCH_FILE are detected

**Comment by haircommander** [comment]
is @joemiller 's [workaround](https://github.com/kubernetes/kubernetes/issues/24957#issuecomment-632881383) sufficient for this? it doesn't seem there's a path forward for a clean API for this behavior, and there seem to be multple work arounds (sidecars, shared pid namespaces, etc)

/priority backlog

**Comment by zffocussss** [comment]
> @alinbalutoiu You could use a sidecar container that watches the mounted configmap for changes and sends a signal to prometheus. That's the best/closest to native way I'm aware of.
> 
> in the pod spec: ` shareProcessNamespace: true`
> 
> A bash script with while loop around inotify
> 
> ```shell
> #!/bin/bash
> # requires these rpm packages on fedora/centos flavors:
> # - procps-ng
> # - inotify-tools
> # or, requires these deb packages on debian/ubuntu flavors:
> # - procps
> # - inotify-tools
> 
> set -eu -o pipefail
> 
> WATCH_FILE="/configmap-mount/prometheus.config"
> INOTIFYWAIT_OPTS="${INOTIFYWAIT_OPTS:-"-e modify -e delete -e delete_self"}"
> 
> main() {
>   while true; do
>     while [[ ! -f "$WATCH_FILE" ]]; do
>       echo "Waiting for $WATCH_FILE to appear ..."
>       sleep 1
>     done
> 
>     echo "Waiting for '$WATCH_FILE' to be deleted or modified..."
>     # shellcheck disable=SC2086
>     if ! inotifywait $INOTIFYWAIT_OPTS "$WATCH_FILE"; then
>       echo "WARNING: inotifywait exited with code $?"
>     fi
>     # small grace period before sending SIGHUP:
>     sleep 1
> 
>     echo "sending SIGHUP to prometheus"
>     if ! pkill -HUP prometheus; then
>       echo "WARNING: 'pkill' exited with code $?"
>     fi
>   done
> }
> main "$@"
> ```
> 
> inotifywait will block until any changes to the $WATCH_FILE are detected.
> 
> You could also try my go-init-sentinel, which is a pid-1 compliant wrapper that allows for watching 1 or more files and sending signals to child processes when those files change. This does not require a sidecar container. https://github.com/joemiller/go-init-sentinel

it is a useful way for us to control process life inside the application container in the same pod.
May I know if there is a common way(maybe CRD, controller) to manage all the pods in a center container/pod?

**Comment by amalic** [comment]
Came here because I was looking for a solution to have pods restarting when dependent secrets change without complicated scripts that will break in some not so distant future. I don't like the sidecar solution becasue in a microservice architecture this would requre tons of resources (energy) wasted just to watch for config or secret updates.

**Comment by amalic** [comment]
> > Came here because I was looking for a solution to have pods restarting when dependent secrets change without complicated scripts that will break in some not so distant future. I don't like the sidecar solution becasue in a microservice architecture this would requre tons of resources (energy) wasted just to watch for config or secret updates.
> 
> The best solution for microservice architecture may be to teach the service to watch for secrets and restart without delegating this responsibility to infrastructure (Kubernetes), and/or call restart of other service via API. Depending on process signal is kind of old view for this task, use service's API instead.

That's how it is being already done. But it's alwazs a custom implementation. It would save time, resources, and potential bughunting sessions when this could be delivered as an out of the box functionality.

**Comment by sftim** [comment]
You could now use an ephemeral container for this, I think, or from a sidecar, if a Pod uses `shareProcessNamespace: true`. Does that help enough that we could document doing that, and then close the issue?

**Comment by sftim** [comment]
I think if I were making any change here, it'd be to allow a Pod to define _which_ containers can share a process namespace. I might want to use an ephemeral container to HUP the log sidecar, but keep the main app container isolated.

**Comment by adampl** [comment]
Ephemeral container for such a simple thing as sending a signal seems like a overkill to me.

**Comment by amalic** [comment]
Friendly reminder about a key [etcd feature](https://etcd.io/)
- **Watch for changes --> Watch specific keys or directories for changes and react to changes in values**

I think this is the exactly right way to handle this requirement.

**Comment by thockin** [comment]
https://github.com/kubernetes/enhancements/issues/1977 is what I want .  It's somewhat involved, and I would love to hear simplifications if anyone has them.  Mostly it needs an owner.

**Comment by sftim** [comment]
@thockin - how about we close this and signpost anyone following _this_ issue to look at https://github.com/kubernetes/enhancements/issues/1977?

We could even mention this issue in the KEP itself for background reading, or under “alternatives considered”.

**Comment by thockin** [comment]
SGTM - we don't really document the relationship between k/k issues and
KEPs, but I can't see we what we need both at the same time.

/close

**Comment by amalic** [comment]
> @thockin - how about we close this and signpost anyone following _this_ issue to look at [kubernetes/enhancements#1977](https://github.com/kubernetes/enhancements/issues/1977)?
> 
> We could even mention this issue in the KEP itself for background reading, or under “alternatives considered”.

This is awful! 

If you read this thread then you will realize that we asked for a way to automatically restart containers if config maps or secrets change. The idea is that the container running a specific workload is **not aware of running in Kubernetes**. Such a restart should happen without running an extra Sidecar or Opearator.

It's a simple requirement that would make day-2-day operations much easier.  Not sure why this is so hard to understand!

**Comment by thockin** [comment]
This issue is specifically about sending a signal to pods from outside.  Somewhere the idea of configmap changes got mixed in.

I don't have a KEP in my pocket for that topic specifically (https://github.com/kubernetes/enhancements/pull/948 is maybe related?), but if someone wants to discuss a proposal for it, I am open to the topic.

**Comment by artheus** [comment]
It would be quite an easy task to make a small service, that watches k8s events for configmaps, secrets etc. Whenever they change, it will just do a rollout restart of the services involved with that configmap/secret/whatever.
It would be a central service, that would not really take much resources at all.
I could write one, if anyone is interested.

**Comment by thockin** [comment]
https://github.com/kubernetes/enhancements/issues/1977 and https://github.com/kubernetes/enhancements/pull/1995 propose a design for this but need an owner.

**Comment by ivankatliarchuk** [comment]
Hi @thockin . TWhile this isn't a critical feature for me, and we've all developed workarounds, its absence does cause occasional friction and bugs me.

I'm happy to take ownership of it. I just need to understand or ask for initial guidance aka where to start?

**Comment by ivankatliarchuk** [comment]
Hi @thockin . While this isn't a critical feature for me, and we've all developed workarounds, its absence does cause occasional friction and bugs me.

I'm happy to take ownership of it. I just need to understand or ask for initial guidance aka where to start?

**Comment by thockin** [comment]
@ivankatliarchuk oh for real. It's been awhile since this was an active topic, but I still think it's something we should do, but there's a lot of stuff in flight around node right now.  That doesn't mean someone couldn't get started on it. 

I should be transparent, I don't think this is a "small" proposal.

Maybe when we make it through this code freeze, we can discuss and see what the state of it is and if it's really something you're interested in pursuing question

**Comment by ivankatliarchuk** [comment]
Gotcha. This could wait for sure. I'm around
